### PR TITLE
Add Windows Jump Lists support in AWT

### PR DESCRIPTION
**Pull request by Alovchin91**
This PR adds Windows Jump Lists support to AWT and fixes compilation on Visual Studio 2022.

### CODE REVIEW

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    var icon: JumpListItemIcon? = null
+}
+
+/**
+ * Represents an icon that can be assigned to a Jump List item.
+ */
+data class JumpListItemIcon(val path: String, val index: Int) {
+    companion object {
+        internal fun fromParts(path: String?, num: Int): JumpListItemIcon? {
+            if (path.isNullOrBlank()) return null
```
Comment (Alovchin91) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:54]: Should be `isNullOrEmpty()` here for consistency with `JumpListInteropItem.fromInterop()`.

File: skiko/src/awtMain/cpp/windows/JumpList.cc [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        }
+    }
+
+    return NULL;
+}
+
+extern "C"
+{
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1init(JNIEnv* env, jobject obj) {
+        ComPtr<ICustomDestinationList> pcdl;
```
Comment (Alovchin91) [inline skiko/src/awtMain/cpp/windows/JumpList.cc:197]: Oops, forgot `CoInitializeWrapper initialize(COINIT_MULTITHREADED);` 😬

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+import kotlinx.coroutines.suspendCancellableCoroutine
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import org.jetbrains.skiko.hostOs
+
+/**
+ * Provides access to Windows Jump List features
+ */
+object JumpList {
```
Comment (Alovchin91) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:11]: @igordmn I've noticed that everywhere in Skiko where a static class is defined a companion object is used. Should I also do it here?

```kotlin
class JumpList private constructor() {
    companion object {
        fun isSupported() ...
    }
}
```

Usage-wise seems pretty much the same to me.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+/**
+ * Represents optional attributes that can be added to a Jump List item.
+ */
+class JumpListItemAttributes {
+    var description: String? = null
+    var icon: JumpListItemIcon? = null
+}
+
+/**
+ * Represents an icon that can be assigned to a Jump List item.
```
Comment (igordmn) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:49]: Let's add a description what is 'path' (a path to image file?) and 'index' (an index of an icon in icon image file?)

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        }
+    }
+}
+
+/**
+ * Provides functions to build the Jump List.
+ *
+ * This is a wrapper for ICustomDestinationList Windows COM interface.
+ */
+class JumpListBuilder internal constructor() : AutoCloseable {
```
Comment (igordmn) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: We can add items, but I didn't find a way how we can react on clicks on these items. Will it be added in a separate PR?

Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: No, I don't think there can be a single source of truth for reacting to the items. There's no "event" or message for this from what I know; the user must pass the "arguments" (see `JumpListItem`, maybe we should explicitly check that the arguments are not an empty string), and the click will launch a new instance of the app with the arguments. That's essentially the only way to react to it.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 3 above)
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:11]: I think it could _theoretically_ work window-wise **if** the window had a separate AppUserModelID, **but** I believe this is discouraged. We also don't set any AppUserModelID at the moment, so the system assigns one automatically to the process itself.

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 5 above)
Comment (igordmn) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: What will be called when we clicked on it?

File: skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt [hunk 6]
```diff
@@ -295,13 +296,25 @@ fun SkikoProjectContext.createLinkJvmBindings(
...
+                    )
+                }
                 addAll(
                     arrayOf(
                         "/NOLOGO",
                         "/DLL",
                         "Advapi32.lib",
                         "gdi32.lib",
                         "Dwmapi.lib",
+                        "ole32.lib",
```
[the remark this answers is not in the document]
Comment (Alovchin91) [reply skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt:316]: Very ubiquitous if we're talking about Windows Desktop 😄 They've been there since... forever, I think you could find ole32.lib in Windows 9x and Propsys.lib in Windows XP. But since we only support Windows 10+, we could also replace most of them with `WindowsApp.lib` or `OneCoreUAP.lib` (ignore the UAP part, it just means that it includes APIs that are available on Desktop devices).

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 5 above)
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: The app that created it + arguments. See:
https://github.com/JetBrains/skiko/blob/9a126efbab1e75f616ebe14fb1429c94a72336b4/skiko/src/awtMain/cpp/windows/JumpList.cc#L107-L111

Comment (igordmn) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: Understood. I confused it with a task list. The current functionality is for recent files that creates a new process every time.

File: skiko/src/awtMain/cpp/windows/JumpList.cc [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+            }
+        }
+    }
+
+    return NULL;
+}
+
+extern "C"
+{
+    JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1init(JNIEnv* env, jobject obj) {
```
Comment (igordmn) [inline skiko/src/awtMain/cpp/windows/JumpList.cc:196]: CI fails here:
https://github.com/JetBrains/skiko/actions/runs/9554903474/job/26421023172?pr=943

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 5 above)
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:65]: In this case it's the same for us. We do pretty much the same thing as the UWP API, because the system provided "recents" category requires `IShellItem`s, but they have a number of limitations. For example, your app must be a registered handler for the file type, and no app except Explorer can add folders to the recents lists (and that's exactly what we need in Fleet). So we create `IShellLink`s instead which are commonly used for tasks but can also go into custom categories (so yeah, for now to have a Recents list you have to define your own category).

File: skiko/src/awtMain/cpp/windows/JumpList.cc (hunk 7 above)
Comment (Alovchin91) [reply skiko/src/awtMain/cpp/windows/JumpList.cc:196]: Yes, I forgot to add a return value to the COM initialization success check 😬

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 8]
```diff
@@ -0,0 +1,156 @@
+package org.jetbrains.skiko.windows
+
+import kotlinx.coroutines.suspendCancellableCoroutine
```
Comment (SergeevPavel) [inline skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:3]: Are the coroutines actually used here?

Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:3]: Oops, no, they are not anymore

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            builder.commit()
+        }
+        else -> error("Jump List is only supported on Windows")
+    }
+}
+
+/**
+ * Represents an item in the Jump List.
+ */
+data class JumpListItem(val title: String, val arguments: String) {
```
[the remark this answers is not in the document]
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt]: Done

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 4 above)
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:67]: Done

File: skiko/src/awtMain/cpp/windows/JumpList.cc (hunk 7 above)
Comment (Alovchin91) [reply skiko/src/awtMain/cpp/windows/JumpList.cc:196]: Done

File: skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Alovchin91) [reply skiko/src/awtMain/kotlin/org/jetbrains/skiko/windows/JumpList.kt:99]: Done!

File: skiko/src/jvmMain/cpp/common/stubs.cc [hunk 10]
```diff
@@ -79,6 +79,39 @@ JNIEXPORT void JNICALL Java_org_jetbrains_skiko_redrawer_Direct3DRedrawer_initFe
...
+    return 0;
+}
+
+JNIEXPORT void JNICALL Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1dispose(JNIEnv *env, jobject obj, jlong ptr) {
+    skikoUnimplemented("Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1dispose");
+}
+
+JNIEXPORT void JNICALL Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1setAppID(JNIEnv *env, jobject obj, jlong ptr, jstring appID) {
+    skikoUnimplemented("Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1setAppID");
+    return 0;
```
Comment (igordmn) [inline skiko/src/jvmMain/cpp/common/stubs.cc:94]:
```suggestion
```

CI fails [here](https://github.com/JetBrains/skiko/actions/runs/9775698526/job/27217252774?pr=943#step:9:1350)
```
  > /home/runner/work/skiko/skiko/skiko/src/jvmMain/cpp/common/stubs.cc:94:5: error: void function 'Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1setAppID' should not return a value [-Wreturn-type]
  >     return 0;
```

File: samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt [hunk 11]
```diff
@@ -20,5 +22,68 @@ fun main() {
...
+    // Check if Jump Lists are supported
+    if (JumpList.isSupported()) {
+        // Start a Jump List building transaction
+        JumpList.build {
+            // If the process has an explicit AppUserModelID, it should be set for the Jump List
+            // before calling [beginList]
+            // setAppID("org.jetbrains.skiko.SkiaMultiplatformSample.awt")
+
+            // Create a new Jump List
+            beginList()
```
Comment (igordmn) [inline samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt:47]: I have this exception running via Gradle:
```
Exception in thread "AWT-EventQueue-0" java.lang.RuntimeException: Native exception in [Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1beginList], code 2147746301: Object is not connected to server 
Context: Failed to BeginList.
	at org.jetbrains.skiko.windows.JumpListBuilder.jumpList_beginList(Native Method)
	at org.jetbrains.skiko.windows.JumpListBuilder.beginList(JumpList.kt:133)
	at org.jetbrains.skiko.sample.App_awtKt$createWindowsJumpList$1.invoke(App.awt.kt:47)
	at org.jetbrains.skiko.sample.App_awtKt$createWindowsJumpList$1.invoke(App.awt.kt:41)
	at org.jetbrains.skiko.windows.JumpList.build(JumpList.kt:26)
	at org.jetbrains.skiko.sample.App_awtKt.createWindowsJumpList(App.awt.kt:41)
	at org.jetbrains.skiko.sample.App_awtKt.main$lambda$2(App.awt.kt:25)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:771)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:722)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:716)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:741)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)
```
Let's wrap the whole function into
```
        try {
            createWindowsJumpList()
        } catch (e: java.lang.RuntimeException) {
            println("Couldn't build the Jump List, most likely because the sample is run via java.exe. ${e.message}")
        }
```

**Review by igordmn — COMMENTED** [review]
LGTM after fixing 2 comments

File: skiko/src/jvmMain/cpp/common/stubs.cc (hunk 10 above)
Comment (Alovchin91) [reply skiko/src/jvmMain/cpp/common/stubs.cc]: Oops, indeed. Fixed!

File: samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt (hunk 11 above)
Comment (Alovchin91) [reply samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt:47]: Thanks for finding this! Turns out it's not a good idea to uninitialise the COM apartment while the Jump List's COM object is alive in it 😅 Fixed!

**Comment by Alovchin91** [comment]
Here's a patch to test it when running via java.exe:

<details>
<summary>diff</summary>

```diff
diff --git a/samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt b/samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt
index 37eb7ee5..d936f1fe 100644
--- a/samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt
+++ b/samples/SkiaMultiplatformSample/src/awtMain/kotlin/org/jetbrains/skiko/sample/App.awt.kt
@@ -41,7 +41,7 @@ private fun createWindowsJumpList() {
         JumpList.build {
             // If the process has an explicit AppUserModelID, it should be set for the Jump List
             // before calling [beginList]
-            // setAppID("org.jetbrains.skiko.SkiaMultiplatformSample.awt")
+            setAppID("org.jetbrains.skiko.SkiaMultiplatformSample.awt")
 
             // Create a new Jump List
             beginList()
diff --git a/skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt b/skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt
index a4982f6f..ea3af1c1 100644
--- a/skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt
+++ b/skiko/buildSrc/src/main/kotlin/tasks/configuration/JvmTasksConfiguration.kt
@@ -316,6 +316,7 @@ fun SkikoProjectContext.createLinkJvmBindings(
                         "ole32.lib",
                         "Propsys.lib",
                         "shcore.lib",
+                        "Shell32.lib",
                         "user32.lib",
                     )
                 )
diff --git a/skiko/src/awtMain/cpp/windows/JumpList.cc b/skiko/src/awtMain/cpp/windows/JumpList.cc
index 9cf55969..36030f9e 100644
--- a/skiko/src/awtMain/cpp/windows/JumpList.cc
+++ b/skiko/src/awtMain/cpp/windows/JumpList.cc
@@ -3,6 +3,7 @@
 #include <windows.h>
 #include <wrl/client.h>
 #include <shobjidl.h>
+#include <shellapi.h>
 #include <propkey.h>
 #include <propvarutil.h>
 
@@ -216,17 +217,41 @@ bool createObjectArray(JNIEnv* env, jobjectArray jobjArray, IObjectArray **ppoa)
     return true;
 }
 
+HRESULT IPropertyStore_SetValue(IPropertyStore *pps, REFPROPERTYKEY pkey, PCWSTR pszValue)
+{
+    PROPVARIANT var;
+    HRESULT hr = InitPropVariantFromString(pszValue, &var);
+    if (SUCCEEDED(hr)) {
+        hr = pps->SetValue(pkey, var);
+        PropVariantClear(&var);
+    }
+    return hr;
+}
+
 extern "C"
 {
     JNIEXPORT jlong JNICALL Java_org_jetbrains_skiko_windows_JumpListBuilder_jumpList_1init(JNIEnv* env, jobject obj) {
         THROW_IF_FAILED(CoInitializeEx(NULL, COINIT_MULTITHREADED), "Failed to initialize COM apartment.", 0L);
 
+        HWND hwnd = FindWindowW(L"SunAwtFrame", NULL);
+        IPropertyStore *pps;
+        HRESULT hr = SHGetPropertyStoreForWindow(hwnd, IID_PPV_ARGS(&pps));
+        if (SUCCEEDED(hr)) {
+            IPropertyStore_SetValue(pps, PKEY_AppUserModel_ID, L"org.jetbrains.skiko.SkiaMultiplatformSample.awt");
+            pps->Release();
+        }
+
         ComPtr<ICustomDestinationList> pcdl;
         THROW_IF_FAILED(
             CoCreateInstance(CLSID_DestinationList, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&pcdl)),
             "Failed to create CLSID_DestinationList.", 0L
         );
 
+        THROW_IF_FAILED(
+            pcdl->DeleteList(L"org.jetbrains.skiko.SkiaMultiplatformSample.awt"),
+            "Failed to delete an existing Jump List.", 0L
+        );
+
         return toJavaPointer(pcdl.Detach());
     }
 
```
</details>

It finds the sample's window and assigns it an AppUserModelID. The Jump List can then be displayed for a window created by the java.exe process.

**Review by igordmn — APPROVED** [review]
